libraries/AP_Scripting/examples/plane_aerobatics.lua

```diff
@@ -0,0 +1,165 @@
+   RUDDER_GAIN = scr_user1_param:get()
+   local rudder = pitch_deg * RUDDER_GAIN * math.sin(math.rad(roll_deg)) / tconst
+   return pitch_rate, rudder
+end
+
+-- a controller for throttle to account for pitch
+function throttle_controller(tconst)
+   local pitch_rad = ahrs:get_pitch()
+   local thr_ff = scr_user3_param:get()
+   local throttle = TRIM_THROTTLE + math.sin(pitch_rad) * thr_ff
+   return constrain(throttle, 0.0, 100.0)
+end
+
+function do_axial_roll(m)
+   -- constant roll rate axial roll
+   if not running then
+      running = true
+      roll_stage = 0
+      gcs:send_text(0, string.format("Starting roll"))
+   end
+   local roll_rate = m:param3()
+   local throttle = m:param4()
+   local pitch_deg = math.deg(ahrs:get_pitch())
+   local roll_deg = math.deg(ahrs:get_roll())
+   if roll_stage == 0 then
+      if roll_deg > 45 then
+         roll_stage = 1
+      end
+   elseif roll_stage == 1 then
+      if roll_deg > -5 and roll_deg < 5 then
+         running = false
+         -- we're done
+         gcs:send_text(0, string.format("Finished roll r=%.1f p=%.1f", roll_deg, pitch_deg))
+         vehicle:nav_scripting_update(true, throttle, 0, 0, 0)
+         roll_stage = 2
+         return
+      end
+   end
+   if roll_stage < 2 then
+      target_pitch = scr_user2_param:get()
+      pitch_rate, rudder = pitch_controller(target_pitch, PITCH_TCONST)
+      vehicle:nav_scripting_update(false, throttle, roll_rate, pitch_rate, 0)
+      if RUDDER_REVERSED == 1 then
+         rudder = -rudder
+      end
+      local rudder_out = math.floor(RUDDER_TRIM + RUDDER_THROW * rudder)
+      rudder_out = constrain(rudder_out, RUDDER_MIN, RUDDER_MAX)
+      SRV_Channels:set_output_pwm_chan_timeout(RUDDER_CHAN, rudder_out, 50)
+   end
+end
+
+local loop_stage = 0
+
+function do_loop(m)
+   -- do one loop with controllable pitch rate and throttle
+   if not running then
+      running = true
+      loop_stage = 0
+      gcs:send_text(0, string.format("Starting loop"))
+   end
+   local pitch_rate = m:param3()
+   local throttle = throttle_controller()
+   local pitch_deg = math.deg(ahrs:get_pitch())
+   local roll_deg = math.deg(ahrs:get_roll())
+   if loop_stage == 0 then
+      if pitch_deg > 60 then
+         loop_stage = 1
+      end
+   elseif loop_stage == 1 then
+      if math.abs(roll_deg) < 90 and pitch_deg > -5 and pitch_deg < 5 then
+         running = false
+         -- we're done
+         gcs:send_text(0, string.format("Finished loop p=%.1f", pitch_deg))
+         vehicle:nav_scripting_update(true, throttle, 0, 0, 0)
+         loop_stage = 2
+         return
+      end
+   end
+   if loop_stage < 2 then
+      local roll_rate = roll_zero_controller(ROLL_TCONST)
+      vehicle:nav_scripting_update(false, throttle, roll_rate, pitch_rate, 0)
+   end
+end
+
+function update()
+   if vehicle:nav_scripting_active() then
+      local cnum = mission:get_current_nav_index()
```

Comment:
Isn't this a (small) race condition where the active mission command could have been changed since vehicle:nav_scripting_active() was called?  I agree with @IamPete1's comments above that it would be better instead of make arg1, arg2 (and whatever other fields go into NAV_SCRIPT_TIME) via another interface rather than asking the script to go to the mission object.